folly/SocketPortability.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2015 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+// While MSVC may not have Unix sockets, Cygwin does.
+#ifdef _MSC_VER
+# define HAVE_UNIX_SOCKETS 0
+#else
+# define HAVE_UNIX_SOCKETS 1
+#endif
+
+#ifdef _MSC_VER
+
+#ifndef __STDC__
+#define __STDC__ 1
+#include <io.h>
+#undef __STDC__
+#else
+#include <io.h>
+#endif
+
+#include <fcntl.h>
+#include <WinSock2.h>
+#include <WS2tcpip.h>
+#include <ws2def.h>
+#include <ws2ipdef.h>
+#include <MSWSock.h>
+
+#include <folly/Portability.h>
+#include <folly/FilePortability.h>
+
+ // We don't actually support either of these flags
+ // currently.
+#define MSG_DONTWAIT 1
+#define MSG_EOR 0
+struct msghdr {
+  void* msg_name;
+  socklen_t msg_namelen;
+  struct iovec* msg_iov;
+  size_t msg_iovlen;
+  void* msg_control;
+  size_t msg_controllen;
+  int msg_flags;
+};
+
+#define SHUT_RD SD_RECEIVE
+#define SHUT_WR SD_SEND
+#define SHUT_RDWR SD_BOTH
+
+namespace folly { namespace socket_portability {
+  SOCKET fd_to_socket(int fd);
+  int socket_to_fd(SOCKET s);
+
+  // Unless you have a case where you would normally have
+  // to reference the function as being explicitly in the
+  // global scope, then you shouldn't be calling these directly.
+  int accept(int s, struct sockaddr* addr, int* addrlen);
+  int bind(int s, const struct sockaddr* name, int namelen);
+  int connect(int s, const struct sockaddr* name, int namelen);
+  int getpeername(int s, struct sockaddr* name, int* namelen);
+  int getsockname(int s, struct sockaddr* name, int* namelen);
+  int getsockopt(int s, int level, int optname, char* optval, int* optlen);
+  int getsockopt(int s, int level, int optname, void* optval, int* optlen);
+  int inet_aton(const char *cp, struct in_addr *inp);
+  const char* inet_ntop(int af, const void* src, char* dst, socklen_t size);
+  int listen(int s, int backlog);
+  int poll(pollfd* fds, int nfds, int timeout);
+  int recv(int s, char* buf, int len, int flags);
+  int recv(int s, void* buf, int len, int flags);
+  int recvfrom(int s, char* buf, int len, int flags, struct sockaddr* from, int* fromlen);
```

Comment:
Please make all lines 80 chars or less. So this would be something like:

```
int recvfrom(int s, char* buf, int len, int flags, struct sockaddr* from, 
                    int* fromlen);
```

There are a **bunch** of these in this pull request; I won't enumerate them all.